sale_purchase_according_declared_stock/models/sale_order.py

```diff
@@ -0,0 +1,254 @@
+            self, po, supplierinfo=False, quantity=False):
+        """ Returns the values to create the PO line from the current SO line.
+            :param po: record of purchase.order
+            :rtype: dict
+            :param quantity: the quantity to force on the PO line
+        """
+        self.ensure_one()
+        # compute quantity from SO line UoM
+        product_quantity = self.product_uom_qty
+        if quantity:
+            product_quantity = quantity
+
+        purchase_qty_uom = self.product_uom._compute_quantity(
+            product_quantity, self.product_id.uom_po_id)
+
+        fpos = po.fiscal_position_id
+        tax = fpos.map_tax(self.product_id.supplier_taxes_id) \
+            if fpos else self.product_id.supplier_taxes_id
+        if tax:
+            tax = tax.filtered(lambda t: t.company_id.id == self.company_id.id)
+
+        price = 0.0
+        seller = self.product_id._select_seller(
+            partner_id=po.partner_id,
+            quantity=purchase_qty_uom,
+            date=po.date_order and po.date_order.date(),
+            uom_id=self.product_id.uom_po_id
+        )
+        if seller:
+            account_tax = self.env['account.tax']
+            price = account_tax.sudo()._fix_tax_included_price_company(
+                seller.price,
+                self.product_id.supplier_taxes_id,
+                tax,
+                self.company_id)
+            if po.currency_id and seller.currency_id != po.currency_id:
+                price = seller.currency_id.compute(price,po.currency_id)
+
+        return {
+            'name': '[%s] %s - pedido %s - %s' %
+                    (self.product_id.default_code, self.product_id.name,
+                     self.order_id.name, self.order_id.partner_id.name),
+            'product_qty': purchase_qty_uom,
+            'product_id': self.product_id.id,
+            'product_uom': self.product_id.uom_po_id.id,
+            'price_unit': price,
+            'date_planned': fields.Date.from_string(po.date_order)
+                            + relativedelta(days=int(seller.delay)),
+            'taxes_id': [(6, 0, tax.ids)],
+            'order_id': po.id,
+            'sale_line_id': self.id,
+        }
+
+    def _select_distribution_seller(self):
+        self.ensure_one()
+
+        res = []
+        sellers = self.env['res.partner'].search([
+            ('property_stock_supplier.usage', '=', 'internal')
+        ])
+        for seller in sellers:
+            stock_qty = self.env['stock.quant']._get_available_quantity(
+                self.product_id, seller.property_stock_supplier
+            )
+            obj_purchase = self.env['purchase.order.line'].search([
+                ('order_id.partner_id', '=', seller.id),
+                ('order_id.state', '=', 'draft'),
+                ('product_id', '=', self.product_id.id),
+            ])
+            purchase_qty = sum(item.product_qty for item in obj_purchase)
```

Comment:
Suggested change (not tested):
```suggestion
            purchase_qty = sum(obj_purchase.mapped("product_qty"))
```

Reply:
ok